quickwit/quickwit-search/src/leaf.rs

```diff
@@ -446,88 +542,134 @@ pub async fn leaf_search(
-    let (split_search_responses, errors): (
-        Vec<tantivy::Result<LeafSearchResponse>>,
-        Vec<(String, SearchError)>,
-    ) = split_search_results
-        .into_iter()
-        .partition_map(|split_search_res| match split_search_res {
-            Ok(Ok(split_search_resp)) => Either::Left(Ok(split_search_resp)),
-            Ok(Err(err)) => Either::Right(err),
-            Err(e) => {
-                warn!("A leaf_search_single_split panicked");
-                Either::Right(("unknown".to_string(), e.into()))
-            }
-        });
+    // we can't use unwrap_or_clone because mutexes aren't Clone
+    let mut incremental_merge_collector = match Arc::try_unwrap(incremental_merge_collector) {
+        Ok(filter_merger) => filter_merger.into_inner().unwrap(),
+        Err(filter_merger) => filter_merger.lock().unwrap().clone(),
+    };
 
-    // Creates a collector which merges responses into one
-    let merge_collector =
-        make_merge_collector(&request, &searcher_context.get_aggregation_limits())?;
+    for result in split_search_results {
+        // splits that did not panic were already added to the collector
+        if let Err(e) = result {
+            incremental_merge_collector.add_failed_split(SplitSearchError {
+                // we could reasonably add a wrapper to the JoinHandle to give us the
+                // split_id anyway
+                split_id: "unknown".to_string(),
+                error: format!("{}", SearchError::from(e)),
+                retryable_error: true,
+            })
+        }
+    }
 
-    // Merging is a cpu-bound task.
-    // It should be executed by Tokio's blocking threads.
-    let span = info_span!("merge_search_responses");
-    let mut merged_search_response = crate::run_cpu_intensive(move || {
-        let _span_guard = span.enter();
-        merge_collector.merge_fruits(split_search_responses)
-    })
-    .await
-    .context("failed to merge split search responses")??;
+    crate::run_cpu_intensive(|| incremental_merge_collector.finalize().map_err(Into::into))
+        .instrument(info_span!("incremental_merge_finalize"))
+        .await
+        .context("failed to merge split search responses")?
+}
 
-    merged_search_response
-        .failed_splits
-        .extend(errors.into_iter().map(|(split_id, err)| SplitSearchError {
-            split_id,
-            error: format!("{err}"),
-            retryable_error: true,
-        }));
-    Ok(merged_search_response)
+#[allow(clippy::too_many_arguments)]
+fn launch_leaf_single_split(
+    request: SearchRequest,
+    searcher_context: Arc<SearcherContext>,
+    index_storage: Arc<dyn Storage>,
+    doc_mapper: Arc<dyn DocMapper>,
+    split: SplitIdAndFooterOffsets,
+    split_filter: Arc<Mutex<CanSplitDoBetter>>,
+    incremental_merge_collector: Arc<Mutex<IncrementalCollector>>,
+    leaf_split_search_permit: tokio::sync::OwnedSemaphorePermit,
+) -> tokio::task::JoinHandle<()> {
+    tokio::spawn(
```

Comment:
could we have moved the `tokio::spawn` to the caller side?
(and made this function `async fn` with the body in the async move block here.

Reply:
(or were you fighting with something tricky?